packages/starknet-snap/src/executeTxn.ts

```diff
@@ -27,7 +33,15 @@ export async function executeTxn(params: ApiParams) {
       addressIndex,
     } = await getKeysFromAddress(keyDeriver, network, state, senderAddress);
 
-    if (await isUpgradeRequired(network, senderAddress)) {
+    const { upgradeRequired, deployRequired, address } = await getCorrectContractAddress(network, publicKey);
```

Comment:
A pubKey can generate a address in cairo0 and cairo1, hence the call is better to respecting the incoming `senderAddress` , as getCorrectContractAddress is to find the correct address from a pubkey

hence i would suggest adding some more util functions to a more specified checking

```typescript
export const isEthBalanceEmpty = async (network: Network, address: string, maxFee: bigint = constants.ZERO) => {
  const etherErc20TokenAddress =
    network.chainId === ETHER_SEPOLIA_TESTNET.chainId ? ETHER_SEPOLIA_TESTNET.address : ETHER_MAINNET.address;

  return num.toBigInt(
    (await getBalance(address, etherErc20TokenAddress, network)) ?? num.toBigInt(constants.ZERO),
  ) <= maxFee;
};


export const isDeployRequired = async (network: Network, address: string, pubKey: string) => {
  logger.log(`isDeployRequired: address = ${address}`);

  const { address: addressLegacy } = getAccContractAddressAndCallDataLegacy(pubKey);

  try {
    if (address === addressLegacy) {
      await getVersion(address, network);
    }
    return true;
  } catch (err) {
    if (!err.message.includes('Contract not found')) {
      throw err;
    }
    return !isEthBalanceEmpty(network, address)
  }
};
```

then we can combine a check logic into a new utils 

```typescript

class UpgradeRequiredError extends Error {
  constructor(msg:string) {
    super(msg);
  }
}

class DeployRequiredError extends Error {
  constructor(msg:string) {
    super(msg);
  }
}

export const validateAccountRequireUpgradeOrDeploy = async (network: Network, address: string, pubKey: string, displayDialog = true) => {
  if (await isUpgradeRequired(network, address)) {
    throw new UpgradeRequiredError('Upgrade required')
  } else if (!(await isDeployRequired(network, address, pubKey))) {
    throw new DeployRequiredError(`Cairo 0 contract address ${address} balance is not empty, deploy required`)
  }
};
```

and then in ur function call
if it is non get-starknet api, just call validateAccountRequireUpgradeOrDeploy
if it is get-starknet api, then
```typescript
try {
      await validateAccountRequireUpgradeOrDeploy(network, senderAddress, publicKey);
    } catch (e) {
      //we can move this part to a snap utils function
      if (e instanceof DeployRequiredError) {
        await showDeployRequestModal(wallet);
      } 
      if (e instanceof UpgradeRequiredError) {
        await showUpgradeRequestModal(wallet);
      }
      throw e
    }
```




in Addtional
the isEthBalanceEmpty can even share with getCorrectContractAddress


WDYT

Reply:
I like it

Reply:
```
export const isDeployRequired = async (network: Network, address: string, pubKey: string) => {
  logger.log(`isDeployRequired: address = ${address}`);

  const { address: addressLegacy } = getAccContractAddressAndCallDataLegacy(pubKey);

  try {
    if (address === addressLegacy) {
      await getVersion(address, network);
    }
    return false; // CHANGE HERE 
  } catch (err) {
    if (!err.message.includes('Contract not found')) {
      throw err;
    }
    return !isEthBalanceEmpty(network, address)
  }
};
```
In the try block shouldn't we return `false` ? because
In case `await getVersion(address, network);` does not fail, it means that the actual address contract is deployed. ==> deploy is not required.

Reply:
good catch